src/properties/property-effects.html

```diff
@@ -1397,33 +1380,35 @@
        *   notification of change to a shared path).
        * @protected
        */
-      _setPathOrUnmanagedProperty(path, value, fromPath) {
+      _setPendingPropertyExternal(path, value, fromPath) {
         let rootProperty = Polymer.Path.root(Array.isArray(path) ? path[0] : path);
         let hasEffect = this._hasPropertyEffect(rootProperty);
         let isPath = (rootProperty !== path);
-        if (!hasEffect) {
-          Polymer.Path.set(this, path, value);
-        } else if (isPath && !fromPath) {
-          // Dirty check changes being set to a path against the actual object,
-          // since this is the entry point for paths into the system; from here
-          // the only dirty checks are against the `__dataTemp` cache to prevent
-          // duplicate work in the same turn only. Note, if this was a
-          // notification of a change already set to a path (fromPath: true),
-          // we always let the change through and skip the `set` since it was
-          // already dirty checked at the point of entry and the underlying
-          // object has already been updated
-          let old = Polymer.Path.get(this, path);
-          path = /** @type {string} */(Polymer.Path.set(this, path, value));
-          // Use property-accessor's simpler dirty check
-          if (!super._shouldPropertyChange(path, value, old)) {
-            return undefined;
-          }
-        }
         if (hasEffect) {
-          return /** @type {string} */(path);
+          if (isPath) {
+            if (!fromPath) {
+              // Dirty check changes being set to a path against the actual object,
+              // since this is the entry point for paths into the system; from here
+              // the only dirty checks are against the `__dataTemp` cache to prevent
+              // duplicate work in the same turn only. Note, if this was a
+              // notification of a change already set to a path (fromPath: true),
+              // we always let the change through and skip the `set` since it was
+              // already dirty checked at the point of entry and the underlying
+              // object has already been updated
+              let old = Polymer.Path.get(this, path);
+              path = /** @type {string} */ Polymer.Path.set(this, path, value);
+              // Use property-accessor's simpler dirty check
+              if (!path || !super._shouldPropertyChange(path, value, old)) {
+                return false;
+              }
+            }
+            this.__dataHasPaths = true;
+          }
+          return this._setPendingProperty(path, value);
         } else {
-          return undefined;
+          Polymer.Path.set(this, path, value);
```

Comment:
use `isPath` here to gate `set` and otherwise just set the property `this[path] = value`